patches/server/0184-Fix-Citizens.patch

```diff
@@ -0,0 +1,105 @@
+-                if (json == null) json = MultiPaper.readStats(server.storageSource.levelDirectory.path().getFileName().toString(), this.file.getName().split("\\.")[0]);
++                // MultiPaper - Start Fix Citizens
++                UUID uuidFile;
++                try {
++                    uuidFile = UUID.fromString(this.file.getName().split("\\.")[0]);
++                } catch (IllegalArgumentException ignored) {
++                    // not a entity
++                    return;
++                }
++                String json = ExternalPlayer.loadedStats.remove(uuidFile);
++                // MultiPaper - End Fix Citizens
++                if (json == null) json = MultiPaper.readStats(server.storageSource.levelDirectory.path().getFileName().toString(), uuidFile.toString());
+                 if (!json.isEmpty()) {
+                     this.parseLocal(server.getFixerUpper(), json);
+                 }
+@@ -86,7 +95,17 @@ public class ServerStatsCounter extends StatsCounter {
+         try {
+             // MultiPaper start
+             if (true) {
+-                MultiPaper.writeStats(server.storageSource.levelDirectory.path().getFileName().toString(), this.file.getName().split("\\.")[0], this.toJson());
++                // MultiPaper - Start Fix Citizens
++                UUID uuidFile;
++                try {
++                    uuidFile = UUID.fromString(this.file.getName().split("\\.")[0]);
++                } catch (IllegalArgumentException ignored) {
++                    // not a entity
++                    return;
++                }
++                String json = ExternalPlayer.loadedStats.remove(uuidFile);
++                MultiPaper.writeStats(server.storageSource.levelDirectory.path().getFileName().toString(), uuidFile.toString(), this.toJson());
++                // MultiPaper - End Fix Citizens
+                 return;
+             }
+             // MultiPaper end
+diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
+index df163434fc5245d893d23ac375bc5f27d1315d4f..4d2f9662d4e803fccaca39f7356f5e0cfe60ffb8 100644
+--- a/src/main/java/puregero/multipaper/MultiPaper.java
++++ b/src/main/java/puregero/multipaper/MultiPaper.java
+@@ -274,7 +274,7 @@ public class MultiPaper {
+     }
+ 
+     public static boolean isRealPlayer(Entity entity) {
+-        return entity instanceof ServerPlayer && !(entity instanceof ExternalPlayer);
++        return entity instanceof ServerPlayer player && !(entity instanceof ExternalPlayer) && !player.getBukkitEntity().hasMetadata("NPC");
+     }
+ 
+     public static boolean isRealPlayer(org.bukkit.entity.Entity bukkitEntity) {
+@@ -282,7 +282,7 @@ public class MultiPaper {
+     }
+ 
+     public static boolean isExternalPlayer(Entity entity) {
+-        return entity instanceof ExternalPlayer;
++        return entity instanceof ExternalPlayer && !entity.getBukkitEntity().hasMetadata("NPC");
```

Comment:
this shouldn't be hard coded as there is many NPC plugins which could have custom meta data other than "NPC"